player/loadfile.c

```diff
@@ -938,6 +943,86 @@ int mp_add_external_file(struct MPContext *mpctx, char *filename,
     return -1;
 }
 
+// Helpers for sub-auto=closest selection
+static bool is_suffix_token(const char *tkn)
+{
+    int len = (int)strlen(tkn);
+    bool alpha = true;
+    for (int i = 0; i < len; i++)
+        alpha &= isalpha((unsigned char)tkn[i]) != 0;
+    if (alpha && (len == 2 || len == 3))
+        return true;
+    const char *special[] = {"eng","en","es","fr","de","pt","ru","jp","ja","zh","chs","cht","sub","subs","sdh","forced","cc",NULL};
+    for (int i = 0; special[i]; i++)
+        if (strcmp(tkn, special[i]) == 0)
+            return true;
+    return false;
+}
+
+static char *normalize_base_name(void *ta_ctx, const char *path)
+{
+    struct bstr base = bstr0(mp_basename(path));
+    base = bstr_strip_ext(base);
+    char *tmpbuf = talloc_strndup(ta_ctx, base.start, base.len);
+    for (int i = 0; tmpbuf[i]; i++)
+        tmpbuf[i] = tolower((unsigned char)tmpbuf[i]);
+    char **tokens = NULL;
+    int ntok = 0;
+    char *p = tmpbuf;
+    while (*p) {
+        while (*p && !isalnum((unsigned char)*p)) p++;
+        if (!*p) break;
+        char *start = p;
+        while (*p && isalnum((unsigned char)*p)) p++;
+        char save = *p; *p = '\0';
+        MP_TARRAY_APPEND(ta_ctx, tokens, ntok, talloc_strdup(ta_ctx, start));
+        *p = save;
+    }
+    while (ntok > 0 && is_suffix_token(tokens[ntok - 1]))
+        ntok--;
+    char *out = talloc_strdup(ta_ctx, "");
+    for (int i = 0; i < ntok; i++)
+        out = talloc_asprintf_append_buffer(out, "%s", tokens[i]);
+    if (!out[0])
+        out = talloc_strdup(ta_ctx, tmpbuf);
+    return out;
+}
+
+static int levenshtein_dist(const char *a, const char *b)
```

Comment:
I believe it would make more sense for this code to live in the `misc` subdirectory. See for example natural_sort.c.